### PR TITLE
feat/add-login-feature

### PR DESCRIPTION
**Pull request by tolehoai**

### CODE REVIEW

File: src/Boostrap/Container.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+     * @param $parameters
+     * @return array
+     * @throws Exception
+     */
+    protected function resolveDependencies($parameters): array
+    {
+        $dependencies = [];
+
+        foreach ($parameters as $parameter) {
+//            $dependency = $parameter->getClass();
```
Comment (bangdinhnfq) [inline src/Boostrap/Container.php]: Remove comment code

File: src/Boostrap/Container.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            if ($parameter->isDefaultValueAvailable()) {
+                $dependencies[] = $parameter->getDefaultValue();
+                continue;
+            }
+            throw new Exception("Can not resolve dependency {$parameter->name}");
+        }
+
+        return $dependencies;
+    }
+}
```
Comment (bangdinhnfq) [inline src/Boostrap/Container.php]: EOL

File: src/Boostrap/Response.php [hunk 3]
```diff
@@ -2,10 +2,48 @@
 
 namespace Tolehoai\CarForRent\Boostrap;
 
+use http\Client\Curl\User;
```
Comment (bangdinhnfq) [inline src/Boostrap/Response.php]: Wrong import

File: src/Controller/UserController.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        try {
+            $errorMessage = '';
+            if ($this->request->isPost()) {
+                $this->userTransfer->fromArray($this->request->getBody());
+                $this->userValidator->validateUserLogin($this->userTransfer);
+                $isLoginSuccess = $this->userService->login($this->userTransfer);
+                if ($isLoginSuccess) {
+                    View::redirect("/");
+                }
+                $errorMessage = 'The username or password is invalid!';
```
Comment (bangdinhnfq) [inline src/Controller/UserController.php]: Should defined as a constant and centralized it.

File: src/Controller/UserController.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        return View::renderView('login', [
+            'title' => 'Login',
+            'username' => $this->userTransfer->getUsername(),
+            'password' => $this->userTransfer->getPassword(),
+            'error' => $errorMessage,
+        ]);
+    }
+
+
+    public
```
Comment (bangdinhnfq) [inline src/Controller/UserController.php]: Should not format like this

File: src/Model/Session.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Tolehoai\CarForRent\Model;
+
+class Session
+{
+    public $id;
```
Comment (bangdinhnfq) [inline src/Model/Session.php]: Should not public variable

File: src/Model/Session.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * @param mixed $userId
+     */
+    public function setUserId($userId): void
+    {
+        $this->userId = $userId;
+    }
+    public $userId;
+}
```
Comment (bangdinhnfq) [inline src/Model/Session.php]: EOL

File: src/Model/User.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
+<?php
+
+namespace Tolehoai\CarForRent\Model;
+
+class User
+{
+    public $id;
```
Comment (bangdinhnfq) [inline src/Model/User.php:7]: Should not public variable

File: src/Repository/SessionRepository.php [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        $statement = $this->connection->prepare("DELETE FROM session WHERE idsession = '$id' ");
+        $statement->execute();
+    }
+
+    public function deleteAll(): void
+    {
+        $this->connection->exec("DELETE FROM session");
+    }
+
+}
```
Comment (bangdinhnfq) [inline src/Repository/SessionRepository.php]: EOL

File: src/Route/route.php [hunk 10]
```diff
@@ -2,15 +2,39 @@
 
 use Tolehoai\CarForRent\Boostrap\Route;
 use Tolehoai\CarForRent\Controller\SiteController;
+use Tolehoai\CarForRent\Controller\UserController;
+use Tolehoai\CarForRent\Database\DatabaseConnection;
 
-Route::get('/', [new SiteController(), 'home']);
+$connection = DatabaseConnection::getConnection();
+//$userRepository = new UserRepository();
```
Comment (bangdinhnfq) [inline src/Route/route.php]: Remove unused code

File: src/Service/DatabaseService.php [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+use PDO;
+use Tolehoai\CarForRent\Database\DatabaseConnection;
+
+class DatabaseService
+{
+    public function getConnection(): PDO
+    {
+        return DatabaseConnection::getConnection();
+    }
+}
```
Comment (bangdinhnfq) [inline src/Service/DatabaseService.php]: EOL

File: src/Service/SessionService.php [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        $sessionId = $_COOKIE[self::$COOKIE_NAME] ?? '';
+        $session = $this->sessionRepository->findById($sessionId);
+        if($session == null){
+            return null;
+        }
+        return $this->userRepository->findById($session->userId);
+    }
+
+}
```
Comment (bangdinhnfq) [inline src/Service/SessionService.php]: EOL

File: src/Transfer/UserTransfer.php [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+     * @param array $params
+     * @return void
+     */
+    public function fromArray(array $params)
+    {
+        $this->username = $params['username'] ?? null;
+        $this->password = $params['password'] ?? null;
+        return $this;
+    }
+}
```
Comment (bangdinhnfq) [inline src/Transfer/UserTransfer.php]: EOL

File: src/Validator/UserValidator.php [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        if (
+            empty($user->getUsername()) ||
+            empty($user->getPassword())
+        ) {
+            throw new ValidationException("Id or password cannot be empty");
+        }
+        return true;
+    }
+}
```
Comment (bangdinhnfq) [inline src/Validator/UserValidator.php]: EOL

File: tests/Tolehoai/CarForRent/UserTransfer/UserTransferTest.php [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+        ];
+        $user = new UserTransfer();
+        $result = $user->fromArray($params);
+        $expected = new UserTransfer();
+        $expected->setUsername('tolehoai');
+        $expected->setPassword('tolehoai');
+        $this->assertEquals($expected, $result);
+
+    }
+}
```
Comment (bangdinhnfq) [inline tests/Tolehoai/CarForRent/UserTransfer/UserTransferTest.php]: EOL

File: tests/Tolehoai/CarForRent/Validator/UserValidatorTest.php [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+                'params' => [
+                    'userRequest'=> $request3
+                ],
+                'expected'=> true
+            ]
+        ];
+    }
+
+
+}
```
Comment (bangdinhnfq) [inline tests/Tolehoai/CarForRent/Validator/UserValidatorTest.php]: EOL

**Review by bangdinhnfq — CHANGES_REQUESTED** [review]

File: src/Controller/SiteController.php [hunk 17]
```diff
@@ -2,32 +2,20 @@
...
-    }
-
-    public function handleContact()
-    {
-        return "Handle submit data";
+            ]
+        );
     }
-}
+}
```
Comment (bangdinhnfq) [inline src/Controller/SiteController.php]: EOL

File: src/Controller/UserApiController.php [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+            }
+            $accessToken = $this->tokenService->create();
+
+            return $this->response->setResponseDataSuccess([...$userTransformer,'AccessToken'=>$accessToken]);
+        } catch (\Exception $exception) {
+            $errorMessage = $exception->getMessage();
+        }
+
+    }
+}
```
Comment (bangdinhnfq) [inline src/Controller/UserApiController.php]: EOL

File: src/Controller/UserApiController.php [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+            $this->userValidator->validateUserLogin($this->userTransfer);
+            $userTransformer = $this->userService->login($this->userTransfer);
+
+            if (!$userTransformer) {
+                $errorMessage = 'Username or Password Invalid';
+                return $this->response->setResponseDataFailed($errorMessage);
+            }
+            $accessToken = $this->tokenService->create();
+
+            return $this->response->setResponseDataSuccess([...$userTransformer,'AccessToken'=>$accessToken]);
```
Comment (bangdinhnfq) [inline src/Controller/UserApiController.php]: Need to change from `AccessToken` => to `token`

File: src/Service/UserService.php [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+        $this->sessionService = $sessionService;
+        $this->userTransformer = $userTransformer;
+    }
+
+    public function login(UserTransfer $userInput)
+    {
+        $existUser = $this->userRepository->findByUsername($userInput->getUsername());
+        if ($existUser && password_verify($userInput->getPassword(), $existUser->getPassword())) {
+            $this->sessionService->create($userInput->getUsername());
+            $userTransformer = $this->userTransformer->transform($existUser);
```
Comment (bangdinhnfq) [inline src/Service/UserService.php]: If login success should return a User, not a UserTransformer

File: src/Repository/UserRepository.php [hunk 21]
```diff
@@ -52,7 +52,7 @@ public function save(RegisterTransfer $user): array
                     $user->getHashPassword(),
                 ]
             );
-            return ["data" => $user];
+            return $user;
         } catch (PDOException $e) {
             return ["error" => "Error when query"];
```
Comment (khajackie2206) [inline src/Repository/UserRepository.php:57]: error must be in plural (errors)

File: tests/Tolehoai/CarForRent/Controller/LoginControllerTest.php [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+        $userValidatorMock = $this->getMockBuilder(UserValidator::class)->disableOriginalConstructor()->getMock();
+        $userValidatorMock->method('validateUserLogin')->willReturn([]);
+        $userRepositoryMock = $this->getMockBuilder(UserRepository::class)->disableOriginalConstructor()->getMock();
+        $sessionServiceMock = $this->getMockBuilder(SessionService::class)->disableOriginalConstructor()->getMock();
+        $userTransformerMock = $this->getMockBuilder(UserTransformer::class)->disableOriginalConstructor()->getMock();
+        $userService = new LoginService($userRepositoryMock, $sessionServiceMock, $userTransformerMock);
+        $result = $userService->login($userTransfer);
+
+    }
+}
```
Comment (khajackie2206) [inline tests/Tolehoai/CarForRent/Controller/LoginControllerTest.php:32]: EOL

File: tests/Tolehoai/CarForRent/Model/CarTest.php [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+    public function testGetName()
+    {
+        $car = new Car();
+        $expected = 'Toyota';
+        $car->setName($expected);
+        $result = $car->getName();
+        $this->assertEquals($expected, $result);
+    }
+
+}
```
Comment (khajackie2206) [inline tests/Tolehoai/CarForRent/Model/CarTest.php:91]: EOL

File: tests/Tolehoai/CarForRent/Service/RegisterServiceTest.php [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+        $databaseService = new DatabaseService();
+        $userTransformer = new UserTransformer();
+        $userRepository = new UserRepository($databaseService, $userTransformer);
+
+        $registerService = new RegisterService($userRepository, $registerTransfer);
+        $result = $registerService->register($registerTransfer);
+        $expected = ['error' => true, 'message' => 'Exit User!'];
+        $this->assertEquals($result, $expected);
+    }
+}
```
Comment (khajackie2206) [inline tests/Tolehoai/CarForRent/Service/RegisterServiceTest.php:48]: EOL

File: tests/Tolehoai/CarForRent/Transfer/CarTransferTest.php [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+        $expected->fromArray($params);
+        $this->assertEquals($params['name'], $expected->getName());
+        $this->assertEquals($params['brand'], $expected->getBrand());
+        $this->assertEquals($params['color'], $expected->getColor());
+        $this->assertEquals($params['luggage'], $expected->getLuggage());
+        $this->assertEquals($params['doors'], $expected->getDoors());
+        $this->assertEquals($params['passenger'], $expected->getPassenger());
+        $this->assertEquals($params['price'], $expected->getPrice());
+    }
+}
```
Comment (khajackie2206) [inline tests/Tolehoai/CarForRent/Transfer/CarTransferTest.php:115]: EOL

File: tests/Tolehoai/CarForRent/Transfer/CarTransferTest.php [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+        $result = 2;
+        $carTransfer->setPrice($result);
+        $expected = $carTransfer->getPrice();
+        $this->assertEquals($result, $expected);
+    }
+
+    public function testFromArray()
+    {
+        $params = [
+
```
Comment (khajackie2206) [inline tests/Tolehoai/CarForRent/Transfer/CarTransferTest.php:94]: an empty line?

File: tests/Tolehoai/CarForRent/Transfer/CarTransferTest.php [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+            'name' => 'vios',
+            'brand' => 'toyota',
+            'color' => 'white',
+
```
Comment (khajackie2206) [inline tests/Tolehoai/CarForRent/Transfer/CarTransferTest.php:98]: Empty line is also here :)))

File: tests/Tolehoai/CarForRent/Transfer/RegisterTransferTest.php [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+            'confirmPassword'=>'tolehoai',
+        ];
+
+        $registerTransfer = new RegisterTransfer();
+        $registerTransfer->fromArray($params);
+        $this->assertEquals($params['username'], $registerTransfer->getUsername());
+        $this->assertEquals($params['password'], $registerTransfer->getPassword());
+        $this->assertEquals($params['confirmPassword'], $registerTransfer->getConfirmPassword());
+    }
+}
```
Comment (khajackie2206) [inline tests/Tolehoai/CarForRent/Transfer/RegisterTransferTest.php:32]: EOL

File: tests/Tolehoai/CarForRent/Transformer/CarTransformerTest.php [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+            'img'=>'car.jpg',
+            'luggage' => 2,
+            'doors' => 2,
+            'passenger' => 2,
+            'price' => 2
+        ];
+
+        $this->assertEquals($expected, $result);
+    }
+}
```
Comment (khajackie2206) [inline tests/Tolehoai/CarForRent/Transformer/CarTransformerTest.php:40]: EOL

File: tests/Tolehoai/CarForRent/Validator/RegisterValidatorTest.php [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+//        $userRegisterRequest = new UserRegisterRequest();
+//        $userRegisterRequest->setSelf('khaitri','123123','123123');
+//        $userRegisterValidator = new RegisterValidatorTest($userRepositoryMock);
+//        $result = $userRegisterValidator->validateUserRegister($userRegisterRequest);
+//        $expectedError = [
+//            'username'=>'Username already exists!'
+//        ];
+//        $this->assertEquals($expectedError,$result);
+//    }
+}
```
Comment (khajackie2206) [inline tests/Tolehoai/CarForRent/Validator/RegisterValidatorTest.php:38]: EOL

**Pull request merged**